extensions/vault/runtime/src/main/java/io/quarkus/vault/VaultTransitSecretEngine.java

```diff
@@ -0,0 +1,108 @@
+
+    // encrypt
+
+    /**
+     * Encrypt a regular string with a vault key configured in the transit secret engine.
+     * equivalent to:
+     * {@code encrypt(keyName, clearData.getBytes("UTF-8"), Optional.empty());}
+     * <p>
+     * this method is usually used in conjunction with {@link #decrypt(String, String)}
+     * 
+     * @see <a href="https://www.vaultproject.io/api/secret/transit/index.html#encrypt-data">encrypt data documentation</a>
+     * @param keyName name of the transit vault key
+     * @param clearData the string to encrypt
+     * @return cipher text
+     */
+    String encrypt(String keyName, String clearData);
+
+    /**
+     * Encrypt a regular string with a vault key configured in the transit secret engine.
+     * if the key does not exist, and the policy specifies a create capability the key will be lazily created
+     * (i.e. upsert). the key can be further customized by specifying transit encryption-key configuration
+     * properties.
+     * 
+     * @see <a href="https://www.vaultproject.io/api/secret/transit/index.html#encrypt-data">encrypt data documentation</a>
+     * @param keyName name of the transit vault key
+     * @param clearData the data to encrypt
+     * @param transitContext an optional context for key derivation
+     * @return cipher text
+     */
+    String encrypt(String keyName, BinaryData clearData, TransitContext transitContext);
+
+    /**
+     * Encrypt a list of elements. This will return a list of results, each being either a ciphertext or an error.
+     * Each element can specify an optional context, allowing key derivation.
+     * 
+     * @see <a href="https://www.vaultproject.io/api/secret/transit/index.html#encrypt-data">encrypt data documentation</a>
+     * @param keyName name of the transit vault key
+     * @param requests the list of elements to encrypt
+     * @return a list of result objects containing either a ciphertext or an error
+     */
+    List<String> encrypt(String keyName, List<EncryptionRequest> requests);
+
+    // decrypt
+
+    BinaryData decrypt(String keyName, String ciphertext);
+
+    BinaryData decrypt(String keyName, String ciphertext, TransitContext transitContext);
+
+    List<DecryptionResult> decrypt(String keyName, List<DecryptionRequest> requests);
+
+    // rewrap
+
+    String rewrap(String keyName, String ciphertext);
+
+    String rewrap(String keyName, String ciphertext, TransitContext transitContext);
+
+    List<EncryptionResult> rewrap(String keyName, List<RewrappingRequest> requests);
+
+    // sign
+
+    String sign(String keyName, String input);
+
+    Signature sign(String keyName, BinaryData input, TransitContext transitContext);
+
+    List<SigningResult> sign(String keyName, SigningBatchRequest batchRequest);
+
+    // verify
+
+    boolean verify(String keyName, String signature, String input);
+
+    boolean verify(String keyName, Verifiable signature, BinaryData input, TransitContext transitContext);
```

Comment:
What is `Verifiable` ? It should be just a string signature ?

Reply:
it was an interface implemented by `Signature` and soon to exist `Hmac`. you verify hmac and signature the same way. and both are strings. with wrapper types in mind, I wanted to have 2 separate types, but only one set of methods.
for simplicity I converted all signature objects to String, and we will have to double methods for hmac, or not implement hmac at all.
spring had it another way where they provide you with a `VaultSignatureVerificationRequest ` where you can specify both fields. but I do not like it because both fields are exclusive (it is one or the other).
having dropped wrapper types in most of the places, I decided to drop `Signature` and `Verifiable`.

Reply:
You are totally right that dealing with both the signature and the hash mutually exclusive. Lets not add a `hash` support until we have the requests for it, one would typically use hash as a thumbprint to identify things, it is not clear to me what the value would be for Quarkus, if one needs to make a remote call to check the hash is valid, alongside the actual value the digest has been created over :-).

Reply:
it is actually 3 different things:
 - hash: get a hash for an input => this can’t be verified (not implemented yet in extension)
 - hmac: generate a hmac for an input => this can be verified (not implemented yet in extension)
 - signature: sign an input => this can be verified

in vault’s api the verify service applies to both signatures and hmac.
to avoid method proliferation, I tried to solve this by introducing a common interface Verifiable, which would be implemented by hmac and signature wrapper classes.
but I think it makes the extension api less obvious. so when we do hmac, we may have to duplicate the verify methods, such as:

```
boolean verifyHmac(String keyName, String hmac, String input);
boolean verifyHmac(String keyName, String hmac, SigningInput input, TransitContext transitContext);
List<Boolean> verifyHmac(String keyName, List<HmacVerificationRequest> requests);
```